### PR TITLE
Move parsed_coin to cosmos and change parsing approach

### PR DESCRIPTION
**Pull request by ribeirotomas1904**

### CODE REVIEW

File: packages/cosmos/src/parsed_coin.rs [hunk 1]
```diff
@@ -29,37 +27,39 @@ impl From<ParsedCoin> for cosmwasm_std::Coin {
...
-            let denom = captures
-                .get(2)
-                .with_context(|| format!("Could not parse denom: {}", s))?
-                .as_str();
-
-            Ok(ParsedCoin {
-                denom: denom.to_owned(),
-                amount: amount.parse()?,
-            })
+            if s.is_empty() {
```
Comment (snoyberg) [inline packages/cosmos/src/parsed_coin.rs:35]: I know I commented to move away from `anyhow`. But just mentioning a trick here is to use `ensure`:

```rust
anyhow::ensure!(!s.is_empty(), "Input is empty");
```

File: packages/cosmos/src/parsed_coin.rs [hunk 2]
```diff
@@ -29,37 +27,39 @@ impl From<ParsedCoin> for cosmwasm_std::Coin {
...
+            let denom_first_index = s
+                .char_indices()
+                .find(|(_, char)| !char.is_ascii_digit())
+                .map(|(index, _)| index);
+
+            match denom_first_index {
+                None => anyhow::bail!("Denom is missing"),
+                Some(0) => anyhow::bail!("Amount is missing"),
+                Some(denom_first_index) => {
+                    let amount = &s[..denom_first_index];
```
Comment (snoyberg) [inline packages/cosmos/src/parsed_coin.rs:48]: Also possible: [`split_at`](https://doc.rust-lang.org/stable/std/primitive.str.html#method.split_at).

File: packages/cosmos/Cargo.toml [hunk 3]
```diff
@@ -40,6 +40,7 @@ flate2 = { version = "1.0.30", features = ["zlib"] }
 strum = "0.26.2"
 strum_macros = "0.26.2"
 tendermint-proto = "0.35.0"
+anyhow = { version = "1.0.86", features = ["backtrace"] }
```
Comment (snoyberg) [inline packages/cosmos/Cargo.toml:43]: Within the cosmos lib, we always want to use well-structured error messages using thiserror. Please see the error.rs module.

**Review by snoyberg — CHANGES_REQUESTED** [review]

File: packages/cosmos/src/parsed_coin.rs (hunk 1 above)
Comment (ribeirotomas1904) [reply packages/cosmos/src/parsed_coin.rs:35]: very handy!

File: packages/cosmos/src/parsed_coin.rs (hunk 2 above)
Comment (ribeirotomas1904) [reply packages/cosmos/src/parsed_coin.rs:48]: ohhhh, that's exactly what i needed

File: packages/cosmos/Cargo.toml (hunk 3 above)
Comment (ribeirotomas1904) [reply packages/cosmos/Cargo.toml:43]: ack

**Review by snoyberg — APPROVED** [review]

**Pull request merged**